src/System.Management.Automation/engine/parser/Parser.cs

```diff
@@ -6426,16 +6426,126 @@ internal Ast CommandRule(bool forDynamicKeyword)
 
         #region Expressions
 
-        private ExpressionAst ExpressionRule()
+        private ExpressionAst ExpressionRule(bool endNumberOnTernaryOpChars = false)
         {
             // G  expression:
             // G      logical-expression
             // G
             // G  logical-expression:
+            // G      binary-expression
+            // G      ternary-expression
+            // G
+            // G  ternary-expression:
+            // G      binary-expression  new-lines:opt   '?'   new-lines:opt   ternary-expression   new-lines:opt   ':'   new-lines:opt   ternary-expression
+
+            RuntimeHelpers.EnsureSufficientExecutionStack();
+            var oldTokenizerMode = _tokenizer.Mode;
+            try
+            {
+                SetTokenizerMode(TokenizerMode.Expression);
+
+                List<Ast> componentAsts = new List<Ast>();
+                ExpressionAst condition = BinaryExpressionRule(endNumberOnTernaryOpChars);
```

Comment:
This allocation can go unused, you should move it closer to the actual use.

Reply:
It looks like it has at most 1 or 2 values and is only used for error handling. I'd actually avoid the allocation completely.

---
In reply to: [303620348](https://github.com/PowerShell/PowerShell/pull/10161#discussion_r303620348) [](ancestors = 303620348)

Reply:
Agree. I will remove the allocation unless we have to return the `ErrorExpressionAst`.

Reply:
Updated.